src/turbolinksAdapter.ts

```diff
@@ -19,12 +21,33 @@ export class TurbolinksAdapter {
     if (response == null) return
     if (response.redirected != true) return
 
-    const action = 'advance'
+    // const action = 'advance'
 
     // TODO: When should we actually use replace other than when specified??
+    // leastbad says: this should be advance unless data-turbolinks-action="replace"
     //   action = 'replace'
 
-    window.Turbolinks.clearCache()
-    window.Turbolinks.visit(location, { action })
+    // i think that this should only happen if data-ujs-morph has been set, otherwise default to Turbolinks visit
+    this.body(event).then(html => {
+      const template = document.createElement('template')
+      template.innerHTML = String(html).trim()
+      morphdom(document.body, template.content, {childrenOnly: true})
+      
+      window.history.pushState({}, '', response.url)
+      // this is obviously not the right thing to do, but we need to tell mrujs to re-init
+      // this hack is causing InvalidAuthenticityToken exceptions in my app when I try to log out
+      document.dispatchEvent(new CustomEvent('turbolinks:load'))
+    })
+
+    // wherever you handle remote forms, we can use the morphdom trick but not do the pushState if there's an error
+    // so depending on the HTTP error that comes back... I guess 4xx errors, right?
+    // seems like 5xx errors should be handled by the app directly
+
+    // window.Turbolinks.clearCache()
+    // window.Turbolinks.visit(location, { action })
+  }
+
+  async body (event: CustomEvent): Promise<string>  {
+    return await event.detail.response.text()
   }
```

Comment:
I think we need to check if its an HTML response first.

https://github.com/hotwired/turbo/blob/aae03ada8be4b46899330364d712cc1ae57f2400/src/http/fetch_response.ts#L34-L36

I still havent decided if we should pre-marshal data for people or not and what the API would look like.

In the meantime here, I think we could check the response for HTML, and if it is HTML we marshal it, if not we carry on with our lives.